# 🎬 MovieLens Analytics Report

В этом отчёте представлен анализ первых 1000 записей из датасета MovieLens.  
Цель — продемонстрировать работу собственного аналитического модуля `movielens_analysis.py`  
и рассказать интересную историю на основе этих данных.


In [1]:
# Импорт необходимых классов
%timeit from movielens_analysis import Ratings, Movies, Tags, Links
from movielens_analysis import Ratings, Movies, Tags, Links

The slowest run took 22.50 times longer than the fastest. This could mean that an intermediate result is being cached.
929 ns ± 1.47 μs per loop (mean ± std. dev. of 7 runs, 1 loop each)


## 📂 Инициализация классов


In [2]:
%timeit ratings = Ratings("../datasets/ratings.csv", "../datasets/movies.csv",)
%timeit movies = Movies("../datasets/movies.csv")
%timeit tags = Tags("../datasets/tags.csv")
%timeit links = Links("../datasets/links.csv", "../datasets/movies.csv")

ratings = Ratings("../datasets/ratings.csv", "../datasets/movies.csv",)
movies = Movies("../datasets/movies.csv")
tags = Tags("../datasets/tags.csv")
links = Links("../datasets/links.csv", "../datasets/movies.csv")

2.26 ms ± 33.1 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
4.27 ms ± 112 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
1.44 ms ± 67.5 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
5.33 ms ± 133 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


## ⭐ Анализ рейтингов


In [3]:
%timeit movies_ratings = ratings.Movies(ratings.ratings, ratings.movies)
%timeit user = ratings.Users(ratings.ratings)

movies_ratings = ratings.Movies(ratings.ratings, ratings.movies)
user = ratings.Users(ratings.ratings)

121 μs ± 2.83 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
329 ns ± 19 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [4]:
%timeit movies_ratings.dist_by_year()
movies_ratings.dist_by_year()

113 μs ± 2.99 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


{1996: 358,
 1999: 82,
 2000: 296,
 2001: 70,
 2005: 121,
 2006: 4,
 2007: 1,
 2011: 39,
 2015: 29}

1996 год лидирует по количеству оценок (358), что может указывать на популярность фильмов этого года.

2000 год также имеет высокий показатель (296).

После 2005 года количество оценок резко снижается, что может быть связано с уменьшением количества фильмов в выборке или изменением активности пользователей.

In [ ]:
%timeit movies_ratings.dist_by_rating()

movies_ratings.dist_by_rating()

{0.5: 24,
 1.0: 39,
 1.5: 11,
 2.0: 57,
 2.5: 7,
 3.0: 253,
 3.5: 17,
 4.0: 292,
 4.5: 33,
 5.0: 267}

Аудитория доброжелательна — критикует редко, чаще хвалит, но идеальных оценок чуть меньше, чем «почти идеальных».

In [ ]:
%timeit movies_ratings.top_by_num_of_ratings(10)

movies_ratings.top_by_num_of_ratings(10)

{'"Usual Suspects': 4,
 'Pulp Fiction (1994)': 4,
 '"Fugitive': 4,
 "Schindler's List (1993)": 4,
 'Batman (1989)': 4,
 '"Silence of the Lambs': 4,
 'Fargo (1996)': 4,
 'Aladdin (1992)': 4,
 'Beauty and the Beast (1991)': 4,
 'Toy Story (1995)': 3}

In [ ]:
%timeit movies_ratings.top_by_ratings(10, metric='median')

movies_ratings.top_by_ratings(10, metric='median')

{'Bottle Rocket (1996)': 5.0,
 'Canadian Bacon (1995)': 5.0,
 'Star Wars: Episode IV - A New Hope (1977)': 5.0,
 'Tommy Boy (1995)': 5.0,
 'Forrest Gump (1994)': 5.0,
 '"Fugitive': 5.0,
 'Jurassic Park (1993)': 5.0,
 'Tombstone (1993)': 5.0,
 'Dances with Wolves (1990)': 5.0,
 'Pinocchio (1940)': 5.0}

Все фильмы в топе имеют максимальную медианную оценку (5.0), что означает высокую согласованность мнений зрителей.

В списке присутствуют как классические фильмы (Forrest Gump, Jurassic Park), так и менее известные (Bottle Rocket).

Нет современных фильмов (после 2000 года) – возможно, старые фильмы получают более единодушные высокие оценки.

In [ ]:
%timeit movies_ratings.top_by_ratings(10)

movies_ratings.top_by_ratings(10)

{'Bottle Rocket (1996)': 5.0,
 'Canadian Bacon (1995)': 5.0,
 'Star Wars: Episode IV - A New Hope (1977)': 5.0,
 'James and the Giant Peach (1996)': 5.0,
 '"Wizard of Oz': 5.0,
 'Citizen Kane (1941)': 5.0,
 '"Adventures of Robin Hood': 5.0,
 'Mr. Smith Goes to Washington (1939)': 5.0,
 'Winnie the Pooh and the Blustery Day (1968)': 5.0,
 '"Three Caballeros': 5.0}

Средние оценки совпадают с медианными, что говорит о малом разбросе оценок (нет резких выбросов).

В топе преобладают фильмы 90-х и более старые, что может указывать на ностальгический фактор.

In [ ]:
%timeit movies_ratings.top_controversial(10)

movies_ratings.top_controversial(10)

Не существует ключа в таблице movies.csv 2018
Не существует ключа в таблице movies.csv 2090
Не существует ключа в таблице movies.csv 2571
Не существует ключа в таблице movies.csv 1704
Не существует ключа в таблице movies.csv 2959


{'My Fair Lady (1964)': 5.06,
 "Schindler's List (1993)": 3.42,
 'Courage Under Fire (1996)': 3.06,
 '"Usual Suspects': 2.42,
 'Dazed and Confused (1993)': 2.25}

"My Fair Lady" (1964) – самый противоречивый фильм (высокий разброс оценок).

"Schindler’s List" (1993) также вызывает полярные мнения (3.42).

Чем ближе значение к 0, тем более единодушны оценки. Например, "Usual Suspects" (2.42) менее спорный, чем "My Fair Lady".

In [ ]:
%timeit user.dist_by_rating()

user.dist_by_rating()

{1: 232, 2: 29, 3: 39, 4: 216, 5: 44, 6: 314, 7: 126}

Пользователь №6 самый активный (314 оценок), а №2 и №3 – наименее активные.

Пользователи 1 и 4 также оставили много оценок (>200).

In [ ]:
%timeit user.top_by_rating(metric='median')

user.top_by_rating(metric='median')

{1: 5.0, 2: 4.0, 3: 0.5, 4: 4.0, 5: 4.0, 6: 3.0, 7: 4.0}

Пользователь №1 ставит только максимальные оценки (5.0) – возможно, он слишком лоялен.

Пользователь №3 имеет очень низкую медиану (0.5) – вероятно, он крайне критичен.

Остальные пользователи ближе к средним значениям.

In [ ]:
%timeit user.top_by_rating()

user.top_by_rating()

{1: 4.37, 2: 3.95, 3: 2.44, 4: 3.56, 5: 3.64, 6: 3.49, 7: 3.35}

Средние оценки ниже медианных, что означает наличие низких оценок, которые "тянут" среднее вниз.

Пользователь №1 остается самым позитивным (4.37), а №3 – самым строгим (2.44).

In [ ]:
%timeit user.top_controversial(5)

user.top_controversial(5)

{3: 4.26, 4: 1.72, 7: 1.65, 5: 0.96, 6: 0.72}

Пользователь №3 – самый непредсказуемый (4.26), что согласуется с его низкой средней оценкой.

Пользователи №5 и №6 ставят более согласованные оценки (значения ближе к 0).

## 🏷️ Анализ тегов


In [ ]:
%timeit tags.most_words(10)

tags.most_words(10)

{'Something for everyone in this one... saw it without and plan on seeing it with kids!': 16,
 'the catholic church is the most corrupt organization in history': 10,
 'Oscar (Best Music - Original Score)': 6,
 'Everything you want is here': 5,
 'based on a true story': 5,
 'heroine in tight suit': 4,
 'lord of the rings': 4,
 'Guardians of the Galaxy': 4,
 'jay and silent bob': 4,
 'based on a book': 4}

Самый длинный тег по словам в нем 16 слов: 
Something for everyone in this one... saw it without and plan on seeing it with kids!

In [ ]:
%timeit tags.longest(10)

tags.longest(10)

['Something for everyone in this one... saw it without and plan on seeing it with kids!',
 'the catholic church is the most corrupt organization in history',
 'audience intelligence underestimated',
 'Oscar (Best Music - Original Score)',
 'assassin-in-training (scene)',
 'Oscar (Best Cinematography)',
 'political right versus left',
 'representation of children',
 'Everything you want is here',
 'artificial intelligence']

Самый длинный тег по символам: Something for everyone in this one... saw it without and plan on seeing it with kids!

In [ ]:
%timeit tags.most_words_and_longest(10)

tags.most_words_and_longest(10)

['Something for everyone in this one... saw it without and plan on seeing it with kids!',
 'the catholic church is the most corrupt organization in history',
 'Oscar (Best Music - Original Score)',
 'Everything you want is here']

Лидер по количеству символов и слов, тоже здесь оказался. Только топ 4 количество слов и символов. Вошли в пересечение

In [ ]:
%timeit tags.most_popular(10)

tags.most_popular(10)

{'funny': 15,
 'sci-fi': 14,
 'twist ending': 12,
 'dark comedy': 12,
 'atmospheric': 10,
 'superhero': 10,
 'comedy': 10,
 'action': 10,
 'suspense': 10,
 'Leonardo DiCaprio': 9}

Самый популярный тег оказался связан с фильмами со смешными, много кто хочет посмеяться

In [ ]:
%timeit tags.tags_with("funny")

tags.tags_with("funny")

['funny', 'stupid but funny', 'very funny']

По самому популярному тегу. Найдены такие теги, чтобы был не тупой юмор и очень смешной

## 🎞️ Анализ жанров фильмов


## В файле movie.csv представлены данные о фильмах в формате
## id фильма | название (год выпуска) | жанры разделённые '|'

### Будем анализирвоать данные при помощи написанных методов

### 1) количество фильмов в каждый год

In [ ]:
%timeit movies.dist_by_release()

movies.dist_by_release()

{1995: 224,
 1994: 184,
 1996: 181,
 1993: 101,
 1992: 23,
 1990: 15,
 1991: 15,
 1989: 14,
 1986: 9,
 1982: 8,
 1940: 8,
 1957: 8,
 1987: 8,
 1980: 8,
 1981: 7,
 1988: 7,
 1979: 7,
 1955: 6,
 1959: 6,
 1968: 6,
 1997: 6,
 1939: 6,
 1985: 6,
 1967: 5,
 1965: 5,
 1951: 5,
 1958: 5,
 1944: 5,
 1941: 5,
 1975: 5,
 1971: 5,
 1984: 5,
 1964: 4,
 1973: 4,
 1954: 4,
 1934: 4,
 1960: 4,
 1963: 4,
 1950: 4,
 1974: 4,
 1983: 4,
 1977: 3,
 1937: 3,
 1972: 3,
 1952: 3,
 1961: 3,
 1953: 3,
 1946: 3,
 1938: 3,
 1956: 3,
 1962: 3,
 1976: 2,
 1969: 2,
 1970: 2,
 1942: 2,
 1945: 2,
 1947: 2,
 1935: 2,
 1936: 2,
 1949: 2,
 1978: 2,
 1943: 1,
 1932: 1,
 1966: 1,
 1948: 1,
 1933: 1,
 1931: 1}

Фильмы, которые будем рассматривать будут примерно 1990-1996 года выпуска. 

Конечно есть некоторые отклонения, но это единичные примеры.

### 2) Количество жанров 

In [ ]:
%timeit movies.dist_by_genres()

movies.dist_by_genres()

{'Drama': 507,
 'Comedy': 365,
 'Romance': 208,
 'Thriller': 179,
 'Action': 158,
 'Adventure': 126,
 'Crime': 122,
 'Children': 100,
 'Fantasy': 69,
 'Sci-Fi': 69,
 'Mystery': 58,
 'Musical': 53,
 'Horror': 51,
 'War': 48,
 'Animation': 37,
 'Documentary': 25,
 'Western': 23,
 'Film-Noir': 18,
 'IMAX': 3}

Самым популярным жанром с большим отрывом является драма. На втором месте идёт комедия. 

Самые непопулярные жанры - фильмы в стиле нуар, вестерны и документальные.

### 3) Фильмы с наибольшим количеством жанров

In [ ]:
%timeit movies.most_genres(20)

movies.most_genres(20)

{'Strange Days (1995)': 6,
 'Lion King, The (1994)': 6,
 'Getaway, The (1994)': 6,
 'Super Mario Bros. (1993)': 6,
 'Beauty and the Beast (1991)': 6,
 'All Dogs Go to Heaven 2 (1996)': 6,
 'Space Jam (1996)': 6,
 'Aladdin and the King of Thieves (1996)': 6,
 'Toy Story (1995)': 5,
 'Money Train (1995)': 5,
 'Copycat (1995)': 5,
 'City of Lost Children, The (Cité des enfants perdus, La) (1995)': 5,
 'Pocahontas (1995)': 5,
 'Bad Boys (1995)': 5,
 "Kid in King Arthur's Court, A (1995)": 5,
 'True Lies (1994)': 5,
 'RoboCop 3 (1993)': 5,
 'Pagemaster, The (1994)': 5,
 'Ghost (1990)': 5,
 'Aladdin (1992)': 5}

У одного фильма количество жанров может достигать 6.

Это существенно, если будет стоять задача разделить фильмы на категории по их жанрам. Брать обособленно один жанр и пытать по нему делить так себе затея.

### 4) При помощи дополнительного метода посмотрим самые частые сочетания жанров

In [ ]:
%timeit movies.dist_by_genre_combination()

movies.dist_by_genre_combination()

{'Drama': 142,
 'Comedy': 94,
 'Comedy|Drama': 54,
 'Drama|Romance': 51,
 'Comedy|Romance': 43,
 'Comedy|Drama|Romance': 28,
 'Documentary': 23,
 'Drama|Thriller': 20,
 'Crime|Drama': 19,
 'Drama|War': 17,
 'Children|Comedy': 14,
 'Adventure|Children': 11,
 'Thriller': 11,
 'Action|Sci-Fi|Thriller': 9,
 'Action|Crime|Thriller': 8,
 'Action|Adventure|Thriller': 8,
 'Children|Drama': 8,
 'Comedy|Crime': 8,
 'Action|Thriller': 7,
 'Crime|Drama|Thriller': 7,
 'Horror': 7,
 'Action': 6,
 'Action|Crime|Drama': 6,
 'Action|Adventure|Sci-Fi': 6,
 'Horror|Thriller': 6,
 'Action|Adventure': 6,
 'Comedy|Musical|Romance': 6,
 'Mystery|Thriller': 5,
 'Action|Crime|Drama|Thriller': 5,
 'Adventure|Children|Drama': 5,
 'Drama|Musical': 5,
 'Children|Comedy|Fantasy': 5,
 'Drama|Fantasy|Romance': 5,
 'Action|Adventure|Sci-Fi|Thriller': 5,
 'Crime|Thriller': 5,
 'Adventure|Children|Fantasy': 4,
 'Action|Adventure|Fantasy': 4,
 'Crime|Mystery|Thriller': 4,
 'Adventure|Drama': 4,
 'Drama|Mystery': 4,
 'Adv

По этим данным можно будет эффективнее разделять фильмы по жанрам. 

Ведь драмма с триллером сильно будет отличаться от драмы с комедией. Хотя оба содержат жанр драмы.

## 🔗 Анализ по ссылкам (Links)


## Парсим данные по фильмам с сайта imdb

In [ ]:
%timeit moives_id = links.first_n_movie_id(1000)
%timeit links.get_imdb(moives_id, fileds)

moives_id = links.first_n_movie_id(1000)
fileds = ['Director', 'Budget', 'Cumulative Worldwide Gross', 'Runtime']
links.get_imdb(moives_id, fileds)

Ошибка при обработке фильма 720: Ошибка сети: 404 Client Error: Not Found for url: https://www.imdb.com/title/tt0118114/


[[1301, 'Fred M. Wilcox', 1900000, 0, datetime.timedelta(seconds=5880)],
 [1300, 'Lasse Hallström', None, 8349284, datetime.timedelta(seconds=6060)],
 [1299, 'Roland Joffé', 14400000, 34700291, datetime.timedelta(seconds=8460)],
 [1298, 'Alan Parker', 12000000, 22274148, datetime.timedelta(seconds=5700)],
 [1297,
  'Martha Coolidge',
  8000000,
  12952019,
  datetime.timedelta(seconds=6480)],
 [1296, 'James Ivory', 3000000, 21062065, datetime.timedelta(seconds=7020)],
 [1295,
  'Philip Kaufman',
  17000000,
  10006806,
  datetime.timedelta(seconds=10260)],
 [1293,
  'Richard Attenborough',
  22000000,
  52768419,
  datetime.timedelta(seconds=11460)],
 [1292, 'Hal Ashby', None, 30179435, datetime.timedelta(seconds=7800)],
 [1291,
  'Steven Spielberg',
  48000000,
  474171806,
  datetime.timedelta(seconds=7620)],
 [1290, 'Howard Deutch', None, 18553948, datetime.timedelta(seconds=5700)],
 [1289, 'Godfrey Reggio', None, 1728699, datetime.timedelta(seconds=5160)],
 [1288, 'Rob Reiner', 250

Мы получили данные в виде
movieid | Режисёр | бюджет фильма | мировые сборы | продолжительность 

### 1) Режисёры с наиобольшим количесвтом фильмов

In [ ]:
%timeit links.top_directors(20)

links.top_directors(20)

{'Alfred Hitchcock': 11,
 'Woody Allen': 8,
 'Frank Capra': 6,
 'Stanley Kubrick': 6,
 'Robert Stevenson': 6,
 'Steven Spielberg': 5,
 'Rob Reiner': 5,
 'James Cameron': 5,
 'Martin Scorsese': 5,
 'John Carpenter': 5,
 'Tony Scott': 5,
 'Vincente Minnelli': 5,
 'Billy Wilder': 5,
 'Peter Jackson': 4,
 'John Huston': 4,
 'Kenneth Branagh': 4,
 'Francis Ford Coppola': 4,
 'Krzysztof Kieslowski': 4,
 'Blake Edwards': 4,
 'Oliver Stone': 4}

Максимально у Alfred Hitchcock 11 

Количсетво фильмов потом резко снижается, и рпедположительно средине занчения будут 2-3 фильма.

### 2) Фильмы с наибольшими бюджетами

In [ ]:
%timeit links.most_expensive(20)

links.most_expensive(20)

{'Akira (1988)': 1100000000,
 'Ghost in the Shell (Kôkaku kidôtai) (1995)': 330000000,
 'Waterworld (1995)': 175000000,
 'Germinal (1993)': 164000000,
 'Cold Fever (Á köldum klaka) (1995)': 130000000,
 'True Lies (1994)': 115000000,
 'Terminator 2: Judgment Day (1991)': 102000000,
 'Cyrano de Bergerac (1990)': 100000000,
 'Eraser (1996)': 100000000,
 'Hunchback of Notre Dame, The (1996)': 100000000,
 'Batman Forever (1995)': 100000000,
 'Cutthroat Island (1995)': 98000000,
 'Twister (1996)': 92000000,
 'Judge Dredd (1995)': 90000000,
 'Die Hard: With a Vengeance (1995)': 90000000,
 'Last Action Hero (1993)': 85000000,
 'Big Blue, The (Grand bleu, Le) (1988)': 80000000,
 'Ransom (1996)': 80000000,
 'Daylight (1996)': 80000000,
 'Space Jam (1996)': 80000000}

К данным было сказано, что не все бюджеты указаны в долларах, поэтому у фильма Акира бюджет миллиард.

Однако таких примеров тут не много, так как можно увидеть Терминатор 2, для которого бюдет точно в долларах указан.

Таким образом, можем посмотерть на самые дорогие фильмы из нашей выборки.

### 3) Самые прибыльные фильмы

In [ ]:
%timeit links.most_profitable(20)

links.most_profitable(20)

{'Jurassic Park (1993)': 1041379926,
 'Lion King, The (1994)': 934161373,
 'E.T. the Extra-Terrestrial (1982)': 786807407,
 'Star Wars: Episode IV - A New Hope (1977)': 764398507,
 'Independence Day (a.k.a. ID4) (1996)': 742400891,
 'Forrest Gump (1994)': 623226465,
 'Star Wars: Episode V - The Empire Strikes Back (1980)': 532016086,
 'Ghost (1990)': 483703557,
 'Aladdin (1992)': 476050219,
 'Home Alone (1990)': 458684675,
 'Star Wars: Episode VI - Return of the Jedi (1983)': 449966382,
 'Pretty Woman (1990)': 449406268,
 'Beauty and the Beast (1991)': 426291298,
 'Indiana Jones and the Last Crusade (1989)': 426171806,
 'Mrs. Doubtfire (1993)': 416286195,
 'Terminator 2: Judgment Day (1991)': 415778573,
 'Twister (1996)': 402580615,
 'Dances with Wolves (1990)': 402208848,
 'Gone with the Wind (1939)': 398405193,
 'Mission: Impossible (1996)': 377696391}

Данный набор данных показывает самые успешные фильмы.

Парк юрского периода вообще вышел в плюс на миллирад.

При этом стоит обратить внимание на то, что парк юрского периода не входит в топ 20 самых дорогих фильмов. Данная информация говорит о том, что не обязательно большой бюджет гарантирует успех фильма в прокате.

### 4) Самые долгие фильмы(в минутах)

In [ ]:
%timeit links.longest(20)

links.longest(20)

{'Gone with the Wind (1939)': 238,
 'Once Upon a Time in America (1984)': 229,
 'Lawrence of Arabia (1962)': 227,
 'Ben-Hur (1959)': 212,
 'Godfather: Part II, The (1974)': 202,
 'Giant (1956)': 201,
 "Schindler's List (1993)": 195,
 'Right Stuff, The (1983)': 193,
 'Nixon (1995)': 192,
 'Gandhi (1982)': 191,
 'Wyatt Earp (1994)': 191,
 'Short Cuts (1993)': 188,
 'Deer Hunter, The (1978)': 183,
 'Wonderful, Horrible Life of Leni Riefenstahl, The (Macht der Bilder: Leni Riefenstahl, Die) (1993)': 183,
 'Dances with Wolves (1990)': 181,
 'Good, the Bad and the Ugly, The (Buono, il brutto, il cattivo, Il) (1966)': 178,
 'Braveheart (1995)': 178,
 'Casino (1995)': 178,
 'Around the World in 80 Days (1956)': 175,
 'Godfather, The (1972)': 175}

Можем видеть, что продолжительности фильмов чуть ли не 4 часов достигают.

При этом значения быстро падают до 3 часов.

Можно считать фильма очень долгим, когда продолжительность выше 3 часов.

### 5) Самые дорогие минуты фильма

In [ ]:
%timeit links.top_cost_per_minute(20)

links.top_cost_per_minute(20)

{'Akira (1988)': 8870967.74,
 'Ghost in the Shell (Kôkaku kidôtai) (1995)': 3975903.61,
 'Cold Fever (Á köldum klaka) (1995)': 1566265.06,
 'Waterworld (1995)': 1296296.3,
 'Hunchback of Notre Dame, The (1996)': 1098901.1,
 'Germinal (1993)': 1025000.0,
 'Judge Dredd (1995)': 937500.0,
 'Space Jam (1996)': 909090.91,
 'Eraser (1996)': 869565.22,
 'Batman Forever (1995)': 826446.28,
 'True Lies (1994)': 815602.84,
 'Twister (1996)': 814159.29,
 'Cutthroat Island (1995)': 790322.58,
 'Terminator 2: Judgment Day (1991)': 744525.55,
 'Cyrano de Bergerac (1990)': 729927.01,
 'Mission: Impossible (1996)': 727272.73,
 'Die Hard: With a Vengeance (1995)': 703125.0,
 'Daylight (1996)': 701754.39,
 'Pocahontas (1995)': 679012.35,
 'Last Man Standing (1996)': 663366.34}

Первым опять является акира, что очевидно, так как бюджет не в долларах указывался.

В целом тут присутсвеют те же фильмы, что в списке самых дорогих по бюджету.

### 6) Допольнительный метод показывающий самых успешных режжисёров

In [ ]:
%timeit links.most_successful_directors(20)

links.most_successful_directors(20)

{'Steven Spielberg': 2926446355,
 'Robert Zemeckis': 989279772,
 'Roger Allers': 934161373,
 'James Cameron': 884712634,
 'Roland Emmerich': 883968153,
 'Chris Columbus': 874970870,
 'George Lucas': 764398507,
 'Jan de Bont': 723028760,
 'Gary Trousdale': 651630149,
 'Ron Howard': 575154955,
 'Jerry Zucker': 556303992,
 'Irvin Kershner': 532016086,
 'Tony Scott': 508317111,
 'Ron Clements': 476050219,
 'Chuck Russell': 470878969,
 'Richard Marquand': 449966382,
 'John McTiernan': 444051805,
 'Jonathan Demme': 434421362,
 'Kevin Reynolds': 431712128,
 'Jon Turteltaub': 425949661}

В топ 20 успешных не присутвет Alfred Hitchcock, хотя у него больш всех фильмов в этой выборке. 

То есть на успех режжисёра влияет не количество фильмов, а качество его фильмов.

На первом месте находится стивен спилберг, хотя в вбыорке представлены всего 5 его фильмов.

### 7) Допольнитльный метод, хронометраж фильма по жанрам

In [ ]:
%timeit links.average_runtime_by_genre(20)

links.average_runtime_by_genre(20)

{'War': 131.12,
 'Western': 129.09,
 'Drama': 115.13,
 'Mystery': 114.05,
 'Action': 113.24,
 'Crime': 112.22,
 'Adventure': 111.62,
 'Romance': 111.22,
 'Thriller': 111.13,
 'Sci-Fi': 109.59,
 'Film-Noir': 106.72,
 'IMAX': 104.0,
 'Comedy': 102.56,
 'Horror': 101.43,
 'Musical': 100.23,
 'Documentary': 99.8,
 'Fantasy': 98.68,
 'Children': 92.48,
 'Animation': 78.83}

Фильмы про войну самые долгие в среднем.

Детские и анимационные фильмы самые короткие. 

Касаемо анимации очевидно, если учитывать года фильмов. Слошком дорого и долго было делать анимации в те времена.

## 📌 Заключение

## 📊 Рейтинги

- **Фильмы 90-х** — особенно 1996 и 2000 годов — получили наибольшее количество оценок.
- **Классические фильмы**, такие как *Star Wars* и *Forrest Gump*, имеют максимальные и стабильно высокие оценки
- Некоторые **старые фильмы** (например, *My Fair Lady*) вызывают споры, вероятно, из-за изменения культурных норм с течением времени.

### 👥 Поведение пользователей

- Пользователи делятся на две категории:
  - **Лояльные** — ставят в основном высокие оценки (например, пользователь №1).
  - **Критичные** — склонны к заниженным рейтингам (например, пользователь №3).
- **Самые активные пользователи** (№6, №1, №4) вносят наибольший вклад в статистику и сильно влияют на общие тренды.

---

## 🏷️ Теги

- **Длинные теги** чаще всего отражают развернутые мнения: личные впечатления, критику, аргументированную оценку.
- **Популярные теги** включают:
  - Жанры
  - Юмор
  - Культовые актеры и франшизы

### 🎭 Тематические тренды

- **Юмор** — главный тренд среди тегов, но с важными оттенками:
  - «Глупый» юмор — популярен, но часто получает неоднозначные оценки.
  - «Качественный» юмор — ценится выше и чаще сопровождается высокими рейтингами.

## 🎞️ Анализ жанров фильмов

- Во-первых по годам фильмов понятно, что данные фильмов в оснвоном 1990-1996 годов, так как выборка из первых 1000 строк
- Самые популярный жанр - драма. Потом уже идут комедии, боевики и остальные.
- У каждого фильма прописано до 6 разных жанров. Стоит учитывать этот факт, при попытке разделить данные по жанрам или провдить какие-то оценки.
- Гораздо лучше смотреть на наборы жанров. Это поможет лучше и правильнее оценивать и разделять фильмы, если есть необходиомсть делать это при помощи жанров.

## 🔗 Анализ по ссылкам фильмов

- Получили данные фильмов с сайта imdb
- Больше всего фильмов у Alfred Hitchcock, однако его 11 фильмов принесли намного меньше, чем 5 фильмов Спилберга
- Большие бюджеты фильмов не гарантируют успех в прокате. Самый успешны фильм даже не в топ-20 по величине бюджета.
- Самые длинные фильмы про войну. Самые короткие - анимации (учитывая года фильмов, более чем логично)